tests/scripts/depends.py

```diff
@@ -0,0 +1,557 @@
+        super().__init__(symbols=symbols, commands=commands, exclude=exclude)
+
+class CipherInfo: # pylint: disable=too-few-public-methods
+    """Collect data about cipher.h."""
+    def __init__(self):
+        self.base_symbols = set()
+        with open('include/mbedtls/cipher.h', encoding="utf-8") as fh:
+            for line in fh:
+                m = re.match(r' *MBEDTLS_CIPHER_ID_(\w+),', line)
+                if m and m.group(1) not in ['NONE', 'NULL', '3DES']:
+                    self.base_symbols.add('MBEDTLS_' + m.group(1) + '_C')
+
+class DomainData:
+    """A container for domains and jobs, used to structurize testing."""
+    def config_symbols_matching(self, regexp):
+        """List the mbedtls_config.h settings matching regexp."""
+        return [symbol for symbol in self.all_config_symbols
+                if re.match(regexp, symbol)]
+
+    def __init__(self, options):
+        """Gather data about the library and establish a list of domains to test."""
+        build_command = [options.make_command, 'CFLAGS=-Werror']
+        build_and_test = [build_command, [options.make_command, 'test']]
+        self.all_config_symbols = set(collect_config_symbols(options))
+        # Find hash modules by name.
+        hash_symbols = self.config_symbols_matching(r'MBEDTLS_(MD|RIPEMD|SHA)[0-9]+_C\Z')
+        # Find elliptic curve enabling macros by name.
+        curve_symbols = self.config_symbols_matching(r'MBEDTLS_ECP_DP_\w+_ENABLED\Z')
+        # Find key exchange enabling macros by name.
+        key_exchange_symbols = self.config_symbols_matching(r'MBEDTLS_KEY_EXCHANGE_\w+_ENABLED\Z')
+        # Find cipher IDs (block permutations and stream ciphers --- chaining
+        # and padding modes are exercised separately) information by parsing
+        # cipher.h, as the information is not readily available in mbedtls_config.h.
+
+        cipher_info = CipherInfo()
+        # Find block cipher chaining and padding mode enabling macros by name.
+        cipher_chaining_symbols = self.config_symbols_matching(r'MBEDTLS_CIPHER_MODE_\w+\Z')
+        cipher_padding_symbols = self.config_symbols_matching(r'MBEDTLS_CIPHER_PADDING_\w+\Z')
+        self.domains = {
+            # Cipher IDs, chaining modes and padding modes. Run the test suites.
+            'cipher_id': ExclusiveDomain(cipher_info.base_symbols,
+                                         build_and_test),
+            'cipher_chaining': ExclusiveDomain(cipher_chaining_symbols,
+                                               build_and_test),
+            'cipher_padding': ExclusiveDomain(cipher_padding_symbols,
+                                              build_and_test),
+            # Elliptic curves. Run the test suites.
+            'curves': ExclusiveDomain(curve_symbols, build_and_test),
+            # Hash algorithms. Exclude configurations with only one
+            # hash which is obsolete. Run the test suites. Exclude
```

Comment:
Sorry, I don't understand how the regex does the bit about obsolete hashes.

Reply:
OK, so the idea is that `MBEDTLS_(MD|RIPEMD|SHA1_|SHA256_|SHA512_)` should not be tested as the only turned on define - first three because they're obsolete, last two because they will be tested as a part of SHA224 / SHA384 testing;
`!MBEDTLS_(SHA256_|SHA512_)` is about not testing configurations where only SHA256 or SHA512 will be turned off, as this is also a part of SHA224 / SHA384 testing. 
Would you like to add this sort of clarification to the code?

Reply:
Ah, I think what got me is that the first line groups 3 obsolete hashes with 2 that are not independent. I think I would find it clearer to have three lines: one for the obsolete hashes, one for 256-512 positive, one for 256-512 negative.

Also, perhaps that's just me, but I would also find it a bit clearer to do SHA256 and SHA512, and skip SHA224 and SHA384 than the other way round - by listing the truncated version I'm more likely to remember that it's not independent from the full version and that's why it's handled in a special way.

Reply:
Now that I think of it, we should test SHA512 in an exclusive domain. It doesn't depend on SHA384.